### PR TITLE
Signup Api 

### PR DESCRIPTION
**Pull request by jenniekibiri**
I implemented the user signup endpoint.
Signup has username, email and password. A Confirm Password check if password entries match before the user is registered
Error handling using express-validator to ensure all fields entered are valid 
and finally if the  user has all the right details he/she is signup  and a token is generated otherwise gives an appropriate error

### CODE REVIEW

**Comment by jenniekibiri** [comment]
#1 
### This PR solves the first issue user signup 
The user can signup with username,email and password all entries are sanitized and throws appropriate error if any
 ### How to test manually
In .env file write your database config 
open postman and select POST then copy this url

```

{
  http://localhost:3000/auth/signup
} 
```

write user details in body in json format
```
{
    "username": "jenny",
    "password": "jenny****",
     "confirmPassword":"jenny***",
    "email": "jenny@gmail.com"
   

}
```
[postman_screenshot](https://i.imgur.com/SspQpCt.png)

###problems & questions
 I had issues with with my database connection string so i ended up using variables in an object to connect to the database

File: src/config/config.js [hunk 1]
```diff
@@ -2,19 +2,26 @@ require('dotenv').config();
...
-    url: process.env.DB_URL_DEV,
-    dialect: 'postgres'
+    // url: process.env.DB_URL_DEV,
+    // dialect: 'postgres'
+    username: process.env.user,
+    password: process.env.password,
+    database: process.env.database,
+    host: process.env.host,
+    dialect: 'postgres',
+    port: process.env.dbport,
```
Comment (mazma1) [inline src/config/config.js]: Why did you take out the `url` from the config?

File: src/app.js [hunk 2]
```diff
@@ -1,18 +1,20 @@
 import express from 'express';
+import morgan from 'morgan';
 import indexRouter from './routes/index';
+import authRouter from './routes/auth';
+
+const expressValidator = require('express-validator');
```
Comment (mazma1) [inline src/app.js]: Change this to an ES6 `import` statement

File: src/controllers/auth.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
+/* eslint-disable consistent-return */
```
Comment (mazma1) [inline src/controllers/auth.js:1]: Why is this rule disabled?

File: src/controllers/auth.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+const model = require('../models/index');
```
Comment (mazma1) [inline src/controllers/auth.js]: Use ES6 `import` statements

File: src/controllers/auth.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+exports.signup = async (req, res) => {
+  const {
+    username, email, confirmPassword,
+  } = req.body;
+
+  let { password } = req.body;
```
Comment (mazma1) [inline src/controllers/auth.js:11]: Why was this not destructured with the rest of `req,body` parameters on line 7 - 9?

File: src/controllers/auth.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  if (password !== confirmPassword) {
+    return res.json({ message: "passwords don't match" });
```
Comment (mazma1) [inline src/controllers/auth.js]: Where is the status code for this response?

File: src/controllers/auth.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  await model.User.findOne({
+    where: {
+      email,
+
+    },
+
+  }).then((user) => {
+    if (user) {
+      return res.status(403).json({
+        error: 'email is already taken Login!',
```
Comment (mazma1) [inline src/controllers/auth.js]: Your error message, `email is already taken Login!`. I don't get the "Login!" in it

File: src/controllers/auth.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      password,
+
+    }).then((user) => {
+      const { id } = user;
+      const token = jwt.sign({ id }, process.env.secret);
+      return res.json({
+        token,
+        user: username,
+      });
+    })
```
Comment (mazma1) [inline src/controllers/auth.js]: Did you confirm what this response should look like from the issue description? Also, you missed out the status code

File: src/models/index.js [hunk 9]
```diff
@@ -1,31 +1,31 @@
...
 const env = process.env.NODE_ENV || 'development';
-const config = require(__dirname + '/../config/config.json')[env];
+const config = require(`${__dirname}/../config/config.js`)[env];
+
 const db = {};
 
 let sequelize;
 if (config.use_env_variable) {
-  sequelize = new Sequelize(process.env[config.use_env_variable], config);
+  sequelize = new Sequelize(process.env[config.use_env_variable]);
```
Comment (mazma1) [inline src/models/index.js]: Why did you change this config?

File: src/validator/index.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
+/* eslint-disable consistent-return */
```
Comment (mazma1) [inline src/validator/index.js:1]: Why is this rule disabled?

**Review by mazma1 — CHANGES_REQUESTED** [review]
Hi @jenniekibiri, I enjoyed reviewing this contribution! First of all, you included an ESLint configuration even without it being part of the issue, that's really great of you! 👍🏽

I've left a few questions just to understand why you made some changes, and also pointed out some things you missed.

Still reviewing, trying to get this up on Heroku so I can access the API on there. I can't deploy right now because there are two lock files (`package.lock` and `yarn.lock`) and it is causing. a conflict. To fix this, delete your local `node_modules` folder and `package.lock` (we will work with yarn) file. Do a fresh dependencies installation with yarn (so you have an updated `yarn.lock`) and then push.

File: src/config/config.js (hunk 1 above)
Comment (jenniekibiri) [reply src/config/config.js]: when i use the connection string it throws authentication error, the password and the username is correct

File: src/controllers/auth.js (hunk 7 above)
Comment (jenniekibiri) [reply src/controllers/auth.js]: if the email is taken that only means that a user with that email is already registered hence the message

File: src/controllers/auth.js (hunk 6 above)
Comment (jenniekibiri) [reply src/controllers/auth.js]: I've fixed that

File: src/controllers/auth.js (hunk 5 above)
Comment (jenniekibiri) [reply src/controllers/auth.js:11]: for the username, email and confirmPassword are constants but the password its changing  because am assigning hash password to it

File: src/controllers/auth.js (hunk 4 above)
Comment (jenniekibiri) [reply src/controllers/auth.js]: okay fixed

File: src/app.js (hunk 2 above)
Comment (jenniekibiri) [reply src/app.js]: fixed

File: src/models/index.js (hunk 9 above)
Comment (jenniekibiri) [reply src/models/index.js]: must have forgotten to add it back when debugging my connection string issues

File: src/controllers/auth.js (hunk 8 above)
Comment (jenniekibiri) [reply src/controllers/auth.js]: should return just the token

File: src/config/config.js (hunk 1 above)
Comment (jenniekibiri) [reply src/config/config.js]: I tried debugging that maybe you can?

File: src/validator/index.js (hunk 10 above)
Comment (jenniekibiri) [reply src/validator/index.js:1]: In this file i want to only return the error and with arrow function it expects the function to return a value and am only returning ones i don't want  to add a else statement so i just return the the error and the function expects another return statement

File: src/controllers/auth.js (hunk 3 above)
Comment (jenniekibiri) [reply src/controllers/auth.js:1]: in this file am using several functions which all require a return value and an if block of code  that require to return a value so as to stop multiple calls sent to client

File: src/config/config.js (hunk 1 above)
Comment (mazma1) [reply src/config/config.js]: I don't know why you have that error but I've attached a screenshot of it working on my system.

Your change is equally fine, we can go with it. You have to however update the `README` and `.env.sample` files to reflect this change.  Also include the `secret` variable you added for the jwt signing step as well. 


Screenshot 2020-07-26 at 08 28 55

File: src/controllers/auth.js (hunk 7 above)
Comment (mazma1) [reply src/controllers/auth.js]: Ok, maybe review the message so it's clearer. Example:
"Email is already taken. Choose a different email or proceed to login." or something along that line

File: src/controllers/auth.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+    model.User.create({
+      username,
+      email,
+      password,
+
+    }).then((user) => {
+      const { id } = user;
+      const token = jwt.sign({ id }, process.env.secret);
+      return res.json({
```
Comment (mazma1) [inline src/controllers/auth.js]: A status code is missing here. Also, include the `username` in the signed token (not as part of the response object). I thought it will be useful on a second thought

File: src/controllers/auth.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        token,
+      });
+    })
+
+      .catch((error) => error);
```
Comment (mazma1) [inline src/controllers/auth.js]: This error handler is not actually sending the error to the client.

How to generate an error: delete the `secret` variable from your .`env` file, test the signup route and see if the error message propagates to you.

**Review by mazma1 — COMMENTED** [review]
We are almost there @jenniekibiri. I added some comments after I got the deployment working, just a few changes here and there

File: src/controllers/auth.js (hunk 7 above)
Comment (jenniekibiri) [reply src/controllers/auth.js]: okay fixed

File: src/config/config.js (hunk 1 above)
Comment (jenniekibiri) [reply src/config/config.js]: updated

File: src/controllers/auth.js (hunk 12 above)
Comment (jenniekibiri) [reply src/controllers/auth.js]: yeah it wasn't throwing any error if you delete secret from  `.env`
`.catch((error) => next(error));` adding next fixes that

**Comment by jenniekibiri** [comment]
> We are almost there @jenniekibiri. I added some comments after I got the deployment working, just a few changes here and there

yaaay!!!!! tested it.  it's working, the deployed one

File: src/controllers/auth.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+      email,
+      password,
+
+    }).then((user) => {
+      const { id } = user;
+      const token = jwt.sign({ id }, process.env.secret);
+      return res.status(200).json({
+        token,
+        user: username,
+      });
```
Comment (mazma1) [inline src/controllers/auth.js]: 1. Is `200` the correct code for successfully creating a resource?
2. The username should be encoded into the token (together with the id) and not in the response object. You should still return only the token.

**Review by mazma1 — APPROVED** [review]
This is at a place we can merge now @jenniekibiri 🎉🎉

**Comment by jenniekibiri** [comment]
> This is at a place we can merge now @jenniekibiri 🎉🎉

we did a nice job!!!!
